File: Design-Patterns/Decorator/Cpp/level_decorator.cpp
#include "attributes_decorator.h"

level_decorator::level_decorator(std::shared_ptr<base_attributes> base_attributes, std::shared_ptr<attributes> parent_attributes, int level) : attributes_decorator(base_attributes, parent_attributes)
{
	this->level = level;
}

int level_decorator::get_max_health()
{
	return parent_attributes_->get_max_health() + (level - 1) * 10;
}

int level_decorator::get_max_mana()
{
	return parent_attributes_->get_max_mana() + (level - 1) * 5;
}

int level_decorator::get_attack()
{
	return parent_attributes_->get_attack() + int(base_attributes_->get_attack() * 0.2 * (level - 1));
}

int level_decorator::get_defense()
{
	return parent_attributes_->get_defense() + int(base_attributes_->get_defense() * 0.1 * (level - 1));
}

int level_decorator::get_critical()
{
	return parent_attributes_->get_critical() + int(base_attributes_->get_critical() * 0.01 * (level - 1));
}

File: Design-Patterns/Decorator/CSharp/Attributes/LevelDecorator.cs
using System;

namespace Decorator.Attributes
{
    public class LevelDecorator : AttributesDecorator
    {
        public int Level;

        public override int MaxHealth => ParentAttributes.MaxHealth + (Level - 1) * 10;
        public override int MaxMana => ParentAttributes.MaxMana + (Level - 1) * 5;
        public override int Attack => ParentAttributes.Attack + Convert.ToInt32(BaseAttributes.Attack * 0.2 * (Level - 1));
        public override int Defense => ParentAttributes.Defense + Convert.ToInt32(BaseAttributes.Defense * 0.1 * (Level - 1));
        public override int Critical => ParentAttributes.Critical + Convert.ToInt32(BaseAttributes.Critical * 0.01 * (Level - 1));        

        public LevelDecorator(BaseAttributes baseAttributes, Attributes parentAttributes, int level) : base(baseAttributes, parentAttributes)
        {
            Level = level;
        }
    }
}
